enterprise/server/sociartifactstore/sociartifactstore.go

```diff
@@ -285,59 +289,52 @@ func sociIndexKey(h v1.Hash) (*repb.Digest, error) {
-	exists, err := s.env.GetCache().Contains(ctx, sociIndexNameResourceName)
-	if err != nil {
-		recordOutcome("soci_index_pointer_contains_error")
-		return nil, err
-	}
-	if !exists {
-		recordOutcome("soci_index_pointer_missing")
-		return nil, status.NotFoundErrorf("soci index for image with manifest config %s not found in cache", imageConfigHash.Hex)
-	}
-	bytes, err := s.env.GetCache().Get(ctx, sociIndexNameResourceName)
-	if err != nil {
-		recordOutcome("soci_index_pointer_read_error")
-		return nil, err
-	}
-	sociIndexResourceName, err := digest.ParseDownloadResourceName(string(bytes))
-	if err != nil {
-		recordOutcome("malformed_soci_index_pointer")
-		return nil, err
-	}
-	sociIndexBytes, err := s.env.GetCache().Get(ctx, sociIndexResourceName.ToProto())
-	if err != nil {
-		recordOutcome("soci_index_read_error")
-		return nil, err
-	}
-	ztocDigests, err := getZtocDigests(sociIndexBytes)
+	actionResult, err := s.env.GetActionCacheServer().GetActionResult(
+		ctx,
+		&repb.GetActionResultRequest{ActionDigest: sociIndexCacheKey})
 	if err != nil {
-		recordOutcome("soci_index_parse_error")
+		if status.IsNotFoundError(err) {
+			recordOutcome("action_result_missing")
+		} else {
+			recordOutcome("action_result_error")
+		}
 		return nil, err
 	}
-	for _, ztocDigest := range ztocDigests {
-		exists, err := s.env.GetCache().Contains(ctx,
-			digest.NewResourceName(ztocDigest, "", rspb.CacheType_CAS, repb.DigestFunction_SHA256).ToProto())
-		if err != nil {
-			recordOutcome("ztoc_contains_error")
-			return nil, err
+
+	var sociIndexDigest *repb.Digest
+	var ztocDigests []*repb.Digest
+	for _, outputFile := range actionResult.OutputFiles {
+		if len(outputFile.NodeProperties.Properties) != 1 {
+			recordOutcome("action_result_missing_node_props")
+			return nil, status.InternalErrorf(
+				"malformed action result, expected exactly 1 node property, found %d",
+				len(outputFile.NodeProperties.Properties))
+		}
+		if outputFile.NodeProperties.Properties[0].Name != artifactTypeKey {
+			recordOutcome("action_result_bad_node_prop_key")
+			return nil, status.InternalErrorf(
+				"malformed action result, expected node property with key 'type', was %s",
+				outputFile.NodeProperties.Properties[0].Name)
 		}
-		if !exists {
-			recordOutcome("ztoc_missing")
-			return nil, status.NotFoundErrorf("ztoc %s not found in cache", ztocDigest.Hash)
+		if outputFile.NodeProperties.Properties[0].Value == sociIndexArtifactType {
+			sociIndexDigest = outputFile.Digest
+		} else if outputFile.NodeProperties.Properties[0].Value == ztocArtifactType {
+			ztocDigests = append(ztocDigests, outputFile.Digest)
+		} else {
+			recordOutcome("action_result_bad_node_prop_value")
+			return nil, status.InternalErrorf(
+				"malformed action result, unrecognized node property value: %s",
+				outputFile.NodeProperties.Properties[0].Value)
 		}
 	}
 	recordOutcome("cached")
```

Comment:
add a check here to verify that sociIndexDigest and ztocDigests were populated

Reply:
Added a check for the SOCI index. It's technically valid (but not very useful) for all the ZToCs to be missing if all of the image layers are below the minimum layer size, so I omitted that chekc.